src/mpid/common/shm/mpidu_shm_alloc.c

```diff
@@ -21,17 +21,589 @@
+
+    } else {
+        char serialized_hnd[MPL_SHM_GHND_SZ] = { 0 };
+
+        /* get serialized handle from rank 0 and deserialize it */
+        mpi_errno = MPIR_Bcast_impl(serialized_hnd, MPL_SHM_GHND_SZ, MPI_CHAR, 0,
+                                    shm_comm_ptr, &errflag);
+        MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+        /* empty handler means root fails */
+        if (strlen(serialized_hnd) == 0)
+            goto map_fail;
+
+        mpl_err = MPL_shm_hnd_deserialize(memory->hnd, serialized_hnd, strlen(serialized_hnd));
+        if (mpl_err != MPL_SHM_SUCCESS) {
+            shm_fail_flag = true;
+            goto result_sync;
+        }
+
+        mpl_err = MPL_shm_seg_attach(memory->hnd, memory->segment_len,
+                                     (void **) &memory->base_addr, 0);
+        if (mpl_err != MPL_SHM_SUCCESS) {
+            shm_fail_flag = true;
+            goto result_sync;
+        } else
+            mapped_flag = true;
+
+      result_sync:
+        mpi_errno = MPIR_Allreduce_impl(&shm_fail_flag, &any_shm_fail_flag, 1, MPI_C_BOOL,
+                                        MPI_LOR, shm_comm_ptr, &errflag);
+        MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+        if (any_shm_fail_flag)
+            goto map_fail;
+    }
+
+    *ptr = memory->base_addr;
+
+    memory_node->ptr = *ptr;
+    memory_node->memory = memory;
+    LL_APPEND(memory_head, memory_tail, memory_node);
+
+    MPIR_CHKPMEM_COMMIT();
+
+  fn_exit:
+    return mpi_errno;
+  map_fail:
+    /* clean up shm mapping resource */
+    if (mapped_flag) {
+        mpl_err =
+            MPL_shm_seg_detach(memory->hnd, (void **) &(memory->base_addr), memory->segment_len);
+        MPIR_ERR_CHKANDJUMP(mpl_err, mpi_errno, MPI_ERR_OTHER, "**detach_shar_mem");
+        *ptr = NULL;
+    }
+    mpl_err = MPL_shm_hnd_finalize(&(memory->hnd));
+    MPIR_ERR_CHKANDJUMP(mpl_err, mpi_errno, MPI_ERR_OTHER, "**remove_shar_mem");
+    mpi_errno = MPIDU_ERR_SHMEM_FAIL;
+    MPIR_CHKPMEM_REAP();
+    goto fn_exit;
+  fn_fail:
+    /* --BEGIN ERROR HANDLING-- */
+    if (memory) {
+        MPL_shm_seg_remove(memory->hnd);
+        MPL_shm_hnd_finalize(&(memory->hnd));
+    }
+    MPIR_CHKPMEM_REAP();
+    goto fn_exit;
+    /* --END ERROR HANDLING-- */
+}
+
 int MPIDU_shm_alloc_symm_all(MPIR_Comm * comm_ptr, size_t len, size_t off, void **ptr)
 {
+    int mpi_errno = MPI_SUCCESS;
+
+    *ptr = NULL;
+
+#ifdef USE_SYM_HEAP
+    mpi_errno = shm_alloc_symm_all(comm_ptr, len, off, ptr);
+#else
+    /* always fail, return and let the caller choose another method */
+    mpi_errno = MPIDU_ERR_SYM_SHMEM_FAIL;
+#endif
 
+    return mpi_errno;
 }
 
 int MPIDU_shm_alloc(MPIR_Comm * shm_comm_ptr, size_t len, void **ptr)
 {
+    int mpi_errno = MPI_SUCCESS;
 
+    *ptr = NULL;
+
+    if (len > 0)
+        mpi_errno = (shm_comm_ptr) ? shm_alloc_node(shm_comm_ptr, len, ptr) : MPIDU_ERR_SHMEM_FAIL;
+
```

Comment:
It seems inconsistent that `MPIDU_shm_alloc` can not handle the `single-process` case but `MPIDU_shm_alloc_symm_all` can. Can we return malloc-ed buffer instead of `MPIDU_ERR_SHMEM_FAIL`?

Reply:
My motivation for this is as follows : `MPIDU_shm_alloc_symm_all` takes a generic communicator which can either be an internode or an intranode comm (in case we only have one node); `MPIDU_shm_alloc` always takes a intranode comm so it does not make sense to consider the case of single process as this should not be called in the first place. Hopefully this makes sense. If not I can fix it.

Reply:
It is fine to me. Although it is not called in perf-critical path, you do not need to check `shm_comm_ptr==NULL` in the caller function because `MPIDU_shm_alloc` can return failure for such a case.

Reply:
I can remove the shm_comm_ptr check and replace it with an Assert as if the user passes a NULL pointer it is wrong caller behavior.